Sources/MongoSwift/BSON/Codec.swift

```diff
@@ -333,3 +333,580 @@ private func getBsonValue(_ value: Any) -> BsonValue? {
+        let _decoder = _BsonDecoder(referencing: data, options: self.options)
+        return try type.init(from: _decoder)
+    }
+}
+
+private class _BsonDecoder: Decoder {
+
+    /// The decoder's storage.
+    fileprivate var storage: _BsonDecodingStorage
+
+    /// Options set on the top-level decoder.
+    fileprivate let options: BsonDecoder._Options
+
+    /// The path to the current point in decoding.
+    fileprivate(set) public var codingPath: [CodingKey]
+
+    /// Contextual user-provided information for use during decoding.
+    public var userInfo: [CodingUserInfoKey: Any] {
+        return self.options.userInfo
+    }
+
+    /// Performs the given closure with the given key pushed onto the end of the current coding path.
+    ///
+    /// - parameter key: The key to push. May be nil for unkeyed containers.
+    /// - parameter work: The work to perform with the key in the path.
+    fileprivate func with<T>(pushedKey key: CodingKey, _ work: () throws -> T) rethrows -> T {
+        self.codingPath.append(key)
+        let ret: T = try work()
+        self.codingPath.removeLast()
+        return ret
+    }
+
+    /// Initializes `self` with the given top-level container and options.
+    fileprivate init(referencing container: BsonValue?, at codingPath: [CodingKey] = [], options: BsonDecoder._Options) {
+        self.storage = _BsonDecodingStorage()
+        self.storage.push(container: container)
+        self.codingPath = codingPath
+        self.options = options
+    }
+
+    // Returns the data stored in this decoder as represented in a container keyed by the given key type.
+    public func container<Key>(keyedBy type: Key.Type) throws -> KeyedDecodingContainer<Key> {
+        guard self.storage.topContainer != nil  else {
+            throw DecodingError.valueNotFound(KeyedDecodingContainer<Key>.self,
+                                  DecodingError.Context(codingPath: self.codingPath,
+                                                        debugDescription: "Cannot get keyed decoding container -- found null value instead."))
+        }
+
+        guard let topContainer = self.storage.topContainer as? Document else {
+            throw DecodingError._typeMismatch(at: self.codingPath, expectation: Document.self, reality: self.storage.topContainer)
+        }
+
+        let container = _BsonKeyedDecodingContainer<Key>(referencing: self, wrapping: topContainer)
+        return KeyedDecodingContainer(container)
+    }
+
+    // Returns the data stored in this decoder as represented in a container appropriate for holding a single primitive value.
+    public func singleValueContainer() throws -> SingleValueDecodingContainer {
+        return self
+    }
+
+    // Returns the data stored in this decoder as represented in a container appropriate for holding values with no keys.
+    public func unkeyedContainer() throws -> UnkeyedDecodingContainer {
```

Comment:
I've refrained from commenting below further on this, but it looks like this change could potentially reduce the code by a large margin

Reply:
So I’ve been looking into this a bit more, and I’m not sure we actually want to get rid of these. 

Consider the following struct: 
```swift
struct SimpleStruct: Decodable, Equatable {
    let ints: [Int]
}
```

The way that gets decoded via the synthesized method is basically:
```swift
init(from decoder: Decoder) throws {
    // get top level keyed container
    let container = decoder.container() 
    // decode what's stored under "ints" to an [Int]
    self.ints = try container.decode([Int].self, forKey: "ints") 
}
```

Under the hood, that `container.decode<T: Decodable>(_ type: T.Type, forKey: Key)` call is pushing the value stored under `ints` onto the container stack, and then calling `Array`'s `init(from decoder: Decoder)` method:
```swift
self.storage.push(container: value)
defer { self.storage.popContainer() }
return try T(from: self)
```

 I *think* that `init` ends up being something like: 
```swift
extension Array {
    init(from decoder: Decoder) throws {
        // get top level unkeyed container
        let container = decoder.unkeyedContainer() 

        var data = [/* some Decodable type */]()
    	while !container.isAtEnd {
            self.append(try container.decode(/* some Decodable type */))
	}

    }
}
```

And then `container.decode(Element.Type.self)`, similarly to the previous decode call, ends up pushing each element of the array onto the stack one by one and then calling `Int`’s `init(from decoder: Decoder)` with each of them, which requests a `singleValueContainer`. 

The one weird thing about this that I’m not sure on is that one would *think* the `container.decode(Element.Type.self)` call made in there would get routed to the `decode(_ type: Int.Type)` method instead of  `decode<T: Decodable>(_ type: T.Type)`  when we’re decoding an `[Int]`, which would allow us to just return the value directly rather than pushing onto stack and then retrieving it.  The only thing I can figure there is that the `Array.init(from decoder: Decoder)` method uses generics somehow and just knows elements are `Decodable`  ,but doesn’t know their specific types. 

So anyway, though I’m still not totally sure what’s going on, it appears to me that both `singleValueContainer` and `unkeyedContainer` s actually still end up getting used with some basic struct configurations we probably want to support.